File: src/types/reservations.ts
/**
 * Tipos específicos para el API de Reservations de Track HS
 */

export interface Guest {
  id: number;
  firstName: string;
  lastName: string;
  email: string;
  phone?: string;
  nationality?: string;
  documentType?: string;
  documentNumber?: string;
}

export interface Property {
  id: number;
  name: string;
  address: string;
  city: string;
  country: string;
  propertyType: string;
  bedrooms: number;
  bathrooms: number;
  maxGuests: number;
}

export interface Reservation {
  id: number;
  alternates?: string[];
  currency: string;
  unitId: number;
  isUnitLocked: boolean;
  isUnitAssigned: boolean;
  isUnitTypeLocked: boolean;
  unitTypeId: number;
  arrivalDate: string;
  departureDate: string;
  earlyArrival: boolean;
  lateDeparture: boolean;
  arrivalTime: string;
  departureTime: string;
  nights: number;
  status: 'Hold' | 'Confirmed' | 'Checked Out' | 'Checked In' | 'Cancelled';
  cancelledAt?: string;
  occupants: Occupant[];
  securityDeposit: {
    required: string;
  };
  updatedAt: string;
  createdAt: string;
  bookedAt: string;
  guestBreakdown: GuestBreakdown;
  type: {
    id: number;
    name: string;
  };
  guaranteePolicy: GuaranteePolicy;
  cancellationPolicy: CancellationPolicy;
  paymentPlan: PaymentPlan[];
  rateType: {
    id: number;
    name: string;
    code: string;
  };
  travelInsuranceProducts: TravelInsuranceProduct[];
  _embedded: {
    unit: Unit;
    contact: Contact;
  };
  _links: {
    self: { href: string };
    cancel: { href: string };
  };
}

export interface Occupant {
  typeId: number;
  name: string;
  handle: string;
  quantity: number;
  included: boolean;
  extraQuantity: number;
  ratePerPersonPerStay: string;
  ratePerStay: string;
}

export interface GuestBreakdown {
  grossRent: string;
  guestGrossDisplayRent: string;
  discount: string;
  promoValue: string;
  discountTotal: number;
  netRent: string;
  guestNetDisplayRent: string;
  actualAdr: string;
  guestAdr: string;
  totalGuestFees: string;
  totalRentFees: string;
  totalItemizedFees: string;
  totalTaxFees: string;
  totalServiceFees: string;
  folioCharges: string;
  subtotal: string;
  guestSubtotal: string;
  totalTaxes: string;
  totalGuestTaxes: string;
  total: string;
  grandTotal: string;
  netPayments: string;
  payments: string;
  refunds: string;
  netTransfers: string;
  balance: string;
  rates: Rate[];
  guestFees: GuestFee[];
  taxes: Tax[];
}

export interface Rate {
  date: string;
  rate: string;
  nights: number;
  isQuoted: boolean;
}

export interface GuestFee {
  id: string;
  name: string;
  displayAs: 'itemize' | 'rent' | 'tax' | 'service';
  quantity: string;
  unitValue: string;
  value: string;
}

export interface Tax {
  id: number;
  name: string;
  amount: string;
}

export interface GuaranteePolicy {
  id: number;
  name: string;
  type: 'Hold' | 'Guarantee' | 'FullDeposit';
  hold: {
    limit: number;
  };
}

export interface CancellationPolicy {
  id: number;
  name: string;
  time: string;
  timezone: string;
  breakpoints: CancellationBreakpoint[];
}

export interface CancellationBreakpoint {
  start: number;
  end: number;
  nonRefundable: boolean;
  nonCancelable: boolean;
  penaltyNights: number;
  penaltyPercent: string;
  penaltyFlat: string;
  description: string;
}

export interface PaymentPlan {
  date: string;
  amount: string;
}

export interface TravelInsuranceProduct {
  id: number;
  status: 'optin' | 'funded' | 'cancelled';
  type: 'Travel Insurance' | 'Master Cancel' | 'Damage Deposit';
  provider: string;
  providerId: number;
  amount: string;
}

export interface Unit {
  id: number;
  name: string;
  shortName: string;
  unitCode: string;
  headline?: string;
  shortDescription: string;
  longDescription: string;
  houseRules?: string;
  nodeId: number;
  unitType: {
    id: number;
    name: string;
  };
  lodgingType: {
    id: number;
    name: string;
  };
  directions?: string;
  checkinDetails?: string;
  timezone: string;
  checkinTime: string;
  hasEarlyCheckin: boolean;
  earlyCheckinTime?: string;
  checkoutTime: string;
  hasLateCheckout: boolean;
  lateCheckoutTime?: string;
  minBookingWindow: number;
  maxBookingWindow: number;
  website?: string;
  phone?: string;
  streetAddress: string;
  extendedAddress?: string;
  locality: string;
  region: string;
  postalCode: string;
  country: string;
  longitude: number;
  latitude: number;
  petsFriendly: boolean;
  maxPets: number;
  eventsAllowed: boolean;
  smokingAllowed: boolean;
  childrenAllowed: boolean;
  minimumAgeLimit?: number;
  isAccessible: boolean;
  area?: number;
  floors?: number;
  maxOccupancy: number;
  securityDeposit: string;
  bedrooms: number;
  fullBathrooms: number;
  threeQuarterBathrooms: number;
  halfBathrooms: number;
  bedTypes: BedType[];
  rooms: Room[];
  amenities: Amenity[];
  amenityDescription: string;
  coverImage: string;
  taxId: number;
  localOffice: LocalOffice;
  regulations: Regulation[];
  updated: {
    availability: string;
    content: string;
    pricing: string;
  };
  updatedAt: string;
  createdAt: string;
  isActive: boolean;
  _links: {
    self: { href: string };
  };
}

export interface BedType {
  id: number;
  name: string;
  count: number;
}

export interface Room {
  name: string;
  type: string;
  sleeps: number;
  description: string;
  hasAttachedBathroom: boolean;
  order?: number;
  airbnbType: string;
  marriottType: string;
  homeawayType: string;
  bedTypes: RoomBedType[];
}

export interface RoomBedType {
  id: number;
  name: string;
  count: number;
  airbnbType: string;
  marriottType: string;
  homeawayType: string;
}

export interface Amenity {
  id: number;
  name: string;
  group: {
    id: number;
    name: string;
  };
}

export interface LocalOffice {
  name: string;
  directions: string;
  email: string;
  phone: string;
  latitude: string;
  longitude: string;
  streetAddress: string;
  extendedAddress: string;
  locality: string;
  region: string;
  postalCode: string;
  country: string;
}

export interface Regulation {
  body: string;
  params: string;
}

export interface Contact {
  id: number;
  firstName: string;
  lastName: string;
  name: string;
  primaryEmail: string;
  secondaryEmail?: string;
  homePhone?: string;
  cellPhone?: string;
  workPhone?: string;
  otherPhone?: string;
  fax?: string;
  streetAddress?: string;
  extendedAddress?: string;
  locality?: string;
  region?: string;
  postalCode?: string;
  country: string;
  notes?: string;
  anniversary?: string;
  birthdate?: string;
  noIdentity?: boolean;
  isVip?: boolean;
  isBlacklist?: boolean;
  isDNR?: boolean;
  tags: Tag[];
  references: Reference[];
  custom: Record<string, any>;
  updatedBy: string;
  createdBy: string;
  updatedAt: string;
  createdAt: string;
  isOwnerContact: boolean;
  _links: {
    self: { href: string };
  };
}

export interface Tag {
  id: number;
}

export interface Reference {
  reference: string;
  salesLinkId?: number;
  channelId?: number;
}

// Tipos para búsqueda de reservaciones
export interface SearchReservationsParams {
  // Paginación
  page?: number;
  size?: number;
  scroll?: number | string;
  
  // Ordenamiento
  sortColumn?: 'name' | 'status' | 'altConf' | 'agreementStatus' | 'type' | 'guest' | 'guests' | 'unit' | 'units' | 'checkin' | 'checkout' | 'nights';
  sortDirection?: 'asc' | 'desc';
  
  // Búsqueda
  search?: string;
  tags?: string;
  updatedSince?: string;
  
  // Filtros por ID
  nodeId?: number | number[];
  unitId?: number | number[];
  reservationTypeId?: number | number[];
  contactId?: number | number[];
  travelAgentId?: number | number[];
  campaignId?: number | number[];
  userId?: number | number[];
  unitTypeId?: number | number[];
  rateTypeId?: number | number[];
  
  // Filtros por fechas
  bookedStart?: string;
  bookedEnd?: string;
  arrivalStart?: string;
  arrivalEnd?: string;
  departureStart?: string;
  departureEnd?: string;
  
  // Filtros especiales
  inHouseToday?: 0 | 1;
  status?: 'Hold' | 'Confirmed' | 'Checked Out' | 'Checked In' | 'Cancelled' | string[];
}

export interface SearchReservationsResponse {
  _embedded: {
    reservations: Reservation[];
  };
  page: number;
  page_count: number;
  page_size: number;
  total_items: number;
  _links: {
    self: { href: string };
    first: { href: string };
    last: { href: string };
    next?: { href: string };
    prev?: { href: string };
  };
}

export interface ReservationResponse {
  data: Reservation;
  success: boolean;
  message?: string;
}

export interface ReservationFilters {
  status?: 'Hold' | 'Confirmed' | 'Checked Out' | 'Checked In' | 'Cancelled';
  channel?: string;
  propertyId?: string;
  dateFrom?: string;
  dateTo?: string;
  guestEmail?: string;
  bookingReference?: string;
}

File: src/trackhs_mcp/types/reservations.py
"""
Modelos Pydantic para Reservations de Track HS API
"""

from typing import Optional, Literal, List, Dict, Any, Union
from pydantic import BaseModel, Field
from .base import PaginationParams, SearchParams

class Guest(BaseModel):
    """Modelo de Guest"""
    id: int = Field(..., description="ID único del huésped")
    first_name: str = Field(..., description="Nombre del huésped")
    last_name: str = Field(..., description="Apellido del huésped")
    email: str = Field(..., description="Email del huésped")
    phone: Optional[str] = Field(default=None, description="Teléfono del huésped")
    nationality: Optional[str] = Field(default=None, description="Nacionalidad del huésped")
    document_type: Optional[str] = Field(default=None, description="Tipo de documento")
    document_number: Optional[str] = Field(default=None, description="Número de documento")

class Property(BaseModel):
    """Modelo de Property"""
    id: int = Field(..., description="ID único de la propiedad")
    name: str = Field(..., description="Nombre de la propiedad")
    address: str = Field(..., description="Dirección de la propiedad")
    city: str = Field(..., description="Ciudad")
    country: str = Field(..., description="País")
    property_type: str = Field(..., description="Tipo de propiedad")
    bedrooms: int = Field(..., description="Número de habitaciones")
    bathrooms: int = Field(..., description="Número de baños")
    max_guests: int = Field(..., description="Máximo de huéspedes")

class Occupant(BaseModel):
    """Modelo de Occupant"""
    type_id: int = Field(..., description="ID del tipo de ocupante")
    name: str = Field(..., description="Nombre del ocupante")
    handle: str = Field(..., description="Handle del ocupante")
    quantity: int = Field(..., description="Cantidad")
    included: bool = Field(..., description="Si está incluido")
    extra_quantity: int = Field(..., description="Cantidad extra")
    rate_per_person_per_stay: str = Field(..., description="Tarifa por persona por estadía")
    rate_per_stay: str = Field(..., description="Tarifa por estadía")

class Rate(BaseModel):
    """Modelo de Rate"""
    date: str = Field(..., description="Fecha")
    rate: str = Field(..., description="Tarifa")
    nights: int = Field(..., description="Noches")
    is_quoted: bool = Field(..., description="Si está cotizada")

class GuestFee(BaseModel):
    """Modelo de GuestFee"""
    id: str = Field(..., description="ID de la tarifa")
    name: str = Field(..., description="Nombre de la tarifa")
    display_as: Literal["itemize", "rent", "tax", "service"] = Field(..., description="Cómo mostrar")
    quantity: str = Field(..., description="Cantidad")
    unit_value: str = Field(..., description="Valor unitario")
    value: str = Field(..., description="Valor")

class Tax(BaseModel):
    """Modelo de Tax"""
    id: int = Field(..., description="ID del impuesto")
    name: str = Field(..., description="Nombre del impuesto")
    amount: str = Field(..., description="Monto del impuesto")

class GuestBreakdown(BaseModel):
    """Modelo de GuestBreakdown"""
    gross_rent: str = Field(..., description="Renta bruta")
    guest_gross_display_rent: str = Field(..., description="Renta bruta mostrada al huésped")
    discount: str = Field(..., description="Descuento")
    promo_value: str = Field(..., description="Valor promocional")
    discount_total: int = Field(..., description="Total de descuentos")
    net_rent: str = Field(..., description="Renta neta")
    guest_net_display_rent: str = Field(..., description="Renta neta mostrada al huésped")
    actual_adr: str = Field(..., description="ADR actual")
    guest_adr: str = Field(..., description="ADR del huésped")
    total_guest_fees: str = Field(..., description="Total de tarifas del huésped")
    total_rent_fees: str = Field(..., description="Total de tarifas de renta")
    total_itemized_fees: str = Field(..., description="Total de tarifas detalladas")
    total_tax_fees: str = Field(..., description="Total de tarifas de impuestos")
    total_service_fees: str = Field(..., description="Total de tarifas de servicio")
    folio_charges: str = Field(..., description="Cargos del folio")
    subtotal: str = Field(..., description="Subtotal")
    guest_subtotal: str = Field(..., description="Subtotal del huésped")
    total_taxes: str = Field(..., description="Total de impuestos")
    total_guest_taxes: str = Field(..., description="Total de impuestos del huésped")
    total: str = Field(..., description="Total")
    grand_total: str = Field(..., description="Gran total")
    net_payments: str = Field(..., description="Pagos netos")
    payments: str = Field(..., description="Pagos")
    refunds: str = Field(..., description="Reembolsos")
    net_transfers: str = Field(..., description="Transferencias netas")
    balance: str = Field(..., description="Balance")
    rates: List[Rate] = Field(..., description="Tarifas")
    guest_fees: List[GuestFee] = Field(..., description="Tarifas del huésped")
    taxes: List[Tax] = Field(..., description="Impuestos")

class GuaranteePolicy(BaseModel):
    """Modelo de GuaranteePolicy"""
    id: int = Field(..., description="ID de la política")
    name: str = Field(..., description="Nombre de la política")
    type: Literal["Hold", "Guarantee", "FullDeposit"] = Field(..., description="Tipo de política")
    hold: Dict[str, int] = Field(..., description="Configuración de retención")

class CancellationBreakpoint(BaseModel):
    """Modelo de CancellationBreakpoint"""
    start: int = Field(..., description="Inicio del rango")
    end: int = Field(..., description="Fin del rango")
    non_refundable: bool = Field(..., description="No reembolsable")
    non_cancelable: bool = Field(..., description="No cancelable")
    penalty_nights: int = Field(..., description="Noches de penalización")
    penalty_percent: str = Field(..., description="Porcentaje de penalización")
    penalty_flat: str = Field(..., description="Penalización fija")
    description: str = Field(..., description="Descripción")

class CancellationPolicy(BaseModel):
    """Modelo de CancellationPolicy"""
    id: int = Field(..., description="ID de la política")
    name: str = Field(..., description="Nombre de la política")
    time: str = Field(..., description="Tiempo")
    timezone: str = Field(..., description="Zona horaria")
    breakpoints: List[CancellationBreakpoint] = Field(..., description="Puntos de quiebre")

class PaymentPlan(BaseModel):
    """Modelo de PaymentPlan"""
    date: str = Field(..., description="Fecha")
    amount: str = Field(..., description="Monto")

class TravelInsuranceProduct(BaseModel):
    """Modelo de TravelInsuranceProduct"""
    id: int = Field(..., description="ID del producto")
    status: Literal["optin", "funded", "cancelled"] = Field(..., description="Estado")
    type: Literal["Travel Insurance", "Master Cancel", "Damage Deposit"] = Field(..., description="Tipo")
    provider: str = Field(..., description="Proveedor")
    provider_id: int = Field(..., description="ID del proveedor")
    amount: str = Field(..., description="Monto")

class Reservation(BaseModel):
    """Modelo de Reservation"""
    id: int = Field(..., description="ID único de la reserva")
    alternates: Optional[List[str]] = Field(default=None, description="Alternativas")
    currency: str = Field(..., description="Moneda")
    unit_id: int = Field(..., description="ID de la unidad")
    is_unit_locked: bool = Field(..., description="Si la unidad está bloqueada")
    is_unit_assigned: bool = Field(..., description="Si la unidad está asignada")
    is_unit_type_locked: bool = Field(..., description="Si el tipo de unidad está bloqueado")
    unit_type_id: int = Field(..., description="ID del tipo de unidad")
    arrival_date: str = Field(..., description="Fecha de llegada")
    departure_date: str = Field(..., description="Fecha de salida")
    early_arrival: bool = Field(..., description="Llegada temprana")
    late_departure: bool = Field(..., description="Salida tardía")
    arrival_time: str = Field(..., description="Hora de llegada")
    departure_time: str = Field(..., description="Hora de salida")
    nights: int = Field(..., description="Noches")
    status: Literal["Hold", "Confirmed", "Checked Out", "Checked In", "Cancelled"] = Field(..., description="Estado")
    cancelled_at: Optional[str] = Field(default=None, description="Fecha de cancelación")
    occupants: List[Occupant] = Field(..., description="Ocupantes")
    security_deposit: Dict[str, str] = Field(..., description="Depósito de seguridad")
    updated_at: str = Field(..., description="Fecha de actualización")
    created_at: str = Field(..., description="Fecha de creación")
    booked_at: str = Field(..., description="Fecha de reserva")
    guest_breakdown: GuestBreakdown = Field(..., description="Desglose del huésped")
    type: Dict[str, Union[int, str]] = Field(..., description="Tipo")
    guarantee_policy: GuaranteePolicy = Field(..., description="Política de garantía")
    cancellation_policy: CancellationPolicy = Field(..., description="Política de cancelación")
    payment_plan: List[PaymentPlan] = Field(..., description="Plan de pagos")
    rate_type: Dict[str, Union[int, str]] = Field(..., description="Tipo de tarifa")
    travel_insurance_products: List[TravelInsuranceProduct] = Field(..., description="Productos de seguro de viaje")
    _embedded: Dict[str, Any] = Field(..., description="Datos embebidos")
    _links: Dict[str, Dict[str, str]] = Field(..., description="Enlaces")

class SearchReservationsParams(PaginationParams, SearchParams):
    """Parámetros para buscar reservas"""
    sort_column: Optional[Literal["name", "status", "altConf", "agreementStatus", "type", "guest", "guests", "unit", "units", "checkin", "checkout", "nights"]] = Field(default="name", description="Columna para ordenar")
    sort_direction: Optional[Literal["asc", "desc"]] = Field(default="asc", description="Dirección de ordenamiento")
    tags: Optional[str] = Field(default=None, description="Tags")
    node_id: Optional[Union[int, List[int]]] = Field(default=None, description="ID del nodo")
    unit_id: Optional[Union[int, List[int]]] = Field(default=None, description="ID de la unidad")
    reservation_type_id: Optional[Union[int, List[int]]] = Field(default=None, description="ID del tipo de reserva")
    contact_id: Optional[Union[int, List[int]]] = Field(default=None, description="ID del contacto")
    travel_agent_id: Optional[Union[int, List[int]]] = Field(default=None, description="ID del agente de viajes")
    campaign_id: Optional[Union[int, List[int]]] = Field(default=None, description="ID de la campaña")
    user_id: Optional[Union[int, List[int]]] = Field(default=None, description="ID del usuario")
    unit_type_id: Optional[Union[int, List[int]]] = Field(default=None, description="ID del tipo de unidad")
    rate_type_id: Optional[Union[int, List[int]]] = Field(default=None, description="ID del tipo de tarifa")
    booked_start: Optional[str] = Field(default=None, description="Fecha de inicio de reserva")
    booked_end: Optional[str] = Field(default=None, description="Fecha de fin de reserva")
    arrival_start: Optional[str] = Field(default=None, description="Fecha de inicio de llegada")
    arrival_end: Optional[str] = Field(default=None, description="Fecha de fin de llegada")
    departure_start: Optional[str] = Field(default=None, description="Fecha de inicio de salida")
    departure_end: Optional[str] = Field(default=None, description="Fecha de fin de salida")
    in_house_today: Optional[Literal[0, 1]] = Field(default=None, description="En casa hoy")
    status: Optional[Union[Literal["Hold", "Confirmed", "Checked Out", "Checked In", "Cancelled"], List[str]]] = Field(default=None, description="Estado")
    group_id: Optional[int] = Field(default=None, description="ID del grupo")
    checkin_office_id: Optional[int] = Field(default=None, description="ID de la oficina de check-in")

class SearchReservationsResponse(BaseModel):
    """Respuesta de búsqueda de reservas"""
    _embedded: Dict[str, List[Reservation]] = Field(..., description="Datos embebidos")
    page: int = Field(..., description="Página actual")
    page_count: int = Field(..., description="Total de páginas")
    page_size: int = Field(..., description="Tamaño de página")
    total_items: int = Field(..., description="Total de elementos")
    _links: Dict[str, Dict[str, str]] = Field(..., description="Enlaces")

class GetReservationParams(BaseModel):
    """Parámetros para obtener una reserva específica"""
    reservation_id: int = Field(..., description="ID de la reserva")

class ReservationResponse(BaseModel):
    """Respuesta de una reserva específica"""
    data: Reservation = Field(..., description="Datos de la reserva")
    success: bool = Field(..., description="Si la operación fue exitosa")
    message: Optional[str] = Field(default=None, description="Mensaje adicional")

class ReservationFilters(BaseModel):
    """Filtros para reservas"""
    status: Optional[Literal["Hold", "Confirmed", "Checked Out", "Checked In", "Cancelled"]] = Field(default=None, description="Estado")
    channel: Optional[str] = Field(default=None, description="Canal")
    property_id: Optional[str] = Field(default=None, description="ID de la propiedad")
    date_from: Optional[str] = Field(default=None, description="Fecha desde")
    date_to: Optional[str] = Field(default=None, description="Fecha hasta")
    guest_email: Optional[str] = Field(default=None, description="Email del huésped")
    booking_reference: Optional[str] = Field(default=None, description="Referencia de reserva")
